lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +435,1145 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+    auto &Summary = Properties.at(EntryPC);
+
+    if (Function.Type == FunctionTypeValue::Fake)
+      continue;
+
+    for (auto &Block : Summary.CFG) {
+      for (auto &Edge : Block.Successors) {
+        if (FunctionEdgeType::isCall(Edge->Type)) {
+          auto *CE = llvm::cast<CallEdge>(Edge.get());
+
+          if (CE->Destination.isValid()) {
+            model::Function &Callee = Binary.Functions.at(CE->Destination);
+            CE->Prototype = Callee.Prototype;
+          } else {
+            auto Prototype = buildPrototype(GCBI, Binary, Summary, Block);
+            CE->Prototype = Binary.recordNewType(std::move(Prototype));
+          }
+        }
+      }
+    }
+
+    Function.CFG = Summary.CFG;
+    revng_check(Function.verify(true));
+  }
+}
+
+static std::optional<uint64_t> electFSO(const auto &MaybeReturns) {
+  auto It = std::min_element(MaybeReturns.begin(),
+                             MaybeReturns.end(),
+                             [](const auto &LHS, const auto &RHS) {
+                               return LHS.second < RHS.second;
+                             });
+  if (It == MaybeReturns.end())
+    return {};
+  return It->second;
+}
+
+static UpcastablePointer<model::FunctionEdge>
+makeEdge(MetaAddress Destination, model::FunctionEdgeType::Values Type) {
+  model::FunctionEdge *Result = nullptr;
+  if (model::FunctionEdgeType::isCall(Type))
+    Result = new model::CallEdge(Destination, Type);
+  else
+    Result = new model::FunctionEdge(Destination, Type);
+  return UpcastablePointer<model::FunctionEdge>(Result);
+};
+
+template<class FunctionOracle>
+SortedVector<model::BasicBlock>
+CFEPAnalyzer<FunctionOracle>::collectDirectCFG(llvm::BasicBlock *Entry,
+                                               llvm::Function *F) {
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (auto &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      auto Start = getBasicBlockPC(&BB);
+      MetaAddress End;
+      uint64_t Size;
+
+      bool Stop = false;
+      model::BasicBlock Block{ Start };
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto [CandidateEnd, CandidateSize] = getPC(Current->getTerminator());
+        if (CandidateEnd > End) {
+          End = CandidateEnd;
+          Size = CandidateSize;
+        }
+
+        for (auto *Succ : successors(Current)) {
+          if (GCBI::isJumpTarget(Succ)) {
+            auto Destination = getBasicBlockPC(Succ);
+            auto Edge = makeEdge(Destination,
+                                 model::FunctionEdgeType::DirectBranch);
+            Block.Successors.insert(Edge);
+          } else if (UnexpectedPCCloned == Succ && succ_size(Current) == 1) {
+            Block.Successors.insert(makeEdge(MetaAddress::invalid(),
+                                             model::FunctionEdgeType::LongJmp));
```

Comment:
Are you sure you don't want the following?
```suggestion
          } else if (UnexpectedPCCloned == Succ) {
            revng_assert(succ_size(Current) == 1);
            Block.Successors.insert(makeEdge(MetaAddress::invalid(),
                                             model::FunctionEdgeType::LongJmp));
```

Reply:
Had thought about `revng_assert(succ_size(Current) == 1);`, but have no doubts here: there are cases in which we don't jump only to `unexpectedpc`, and in those cases we don't want to emit any edge.

Reply:
Right, then put a comment stating that if we're jumping *only* to unexpectedpc we need to record this, otherwise unexpectedpc is just a safety net for errors in our analyses that we want to ignore here.